File: frontend/src/types/portfolio.ts
export interface Holding {
  id: number;
  portfolio_id: number;
  asset_type: 'equity' | 'debt' | 'gold' | 'international';
  asset_name: string;
  quantity: number;
  current_price: number;
  purchase_price: number;
  allocation: number;
  ai_recommendation?: 'buy' | 'hold' | 'sell';
  created_at: string;
  updated_at: string;
}

export interface Portfolio {
  id: number;
  user_id: number;
  total_value: number;
  total_gain_loss: number;
  total_gain_loss_percent: number;
  last_updated: string;
}

export interface PortfolioResponse {
  portfolio: Portfolio;
  holdings: Holding[];
  performance: PortfolioPerformance;
}

export interface PortfolioPerformance {
  total_value: number;
  total_gain_loss: number;
  total_gain_loss_percent: number;
  asset_allocation: {
    equity: number;
    debt: number;
    gold: number;
    international: number;
  };
  returns_by_asset: {
    equity: number;
    debt: number;
    gold: number;
    international: number;
  };
}

export interface OptimizePortfolioRequest {
  user_id: number;
  risk_category: 'conservative' | 'moderate' | 'aggressive' | 'very_aggressive';
  investment_amount: number;
  goals?: Array<{
    goal_name: string;
    target_amount: number;
    target_date: string;
  }>;
}

export interface OptimizePortfolioResponse {
  allocation: {
    equity: number;
    debt: number;
    gold: number;
    international: number;
  };
  expected_return: number;
  risk_metrics: {
    volatility: number;
    sharpe_ratio: number;
    max_drawdown: number;
  };
  behavioral_adjustments: Record<string, any>;
}

export interface RebalancePortfolioRequest {
  user_id: number;
  target_allocation: {
    equity: number;
    debt: number;
    gold: number;
    international: number;
  };
}

export interface RebalancePortfolioResponse {
  recommendations: Array<{
    action: 'buy' | 'sell';
    asset_type: string;
    asset_name?: string;
    amount: number;
    reason: string;
  }>;
  current_value: number;
}

export interface AddHoldingRequest {
  portfolio_id: number;
  asset_type: 'equity' | 'debt' | 'gold' | 'international';
  asset_name: string;
  quantity: number;
  current_price: number;
  purchase_price: number;
  allocation: number;
  ai_recommendation?: 'buy' | 'hold' | 'sell';
}

File: Backend/models/portfolio.py
"""
Portfolio and Holdings models
"""

from extensions import db
from datetime import datetime
from decimal import Decimal

class Portfolio(db.Model):
    __tablename__ = 'portfolios'
    
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id', ondelete='CASCADE'), nullable=False, index=True)
    total_value = db.Column(db.Numeric(15, 2), default=Decimal('0.00'), nullable=False)
    total_gain_loss = db.Column(db.Numeric(15, 2), default=Decimal('0.00'), nullable=False)
    total_gain_loss_percent = db.Column(db.Numeric(10, 4), default=Decimal('0.0000'), nullable=False)
    last_updated = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    holdings = db.relationship('Holding', backref='portfolio', lazy='dynamic', cascade='all, delete-orphan')
    
    def to_dict(self):
        """Convert portfolio to dictionary"""
        return {
            'id': self.id,
            'user_id': self.user_id,
            'total_value': float(self.total_value) if self.total_value else 0.0,
            'total_gain_loss': float(self.total_gain_loss) if self.total_gain_loss else 0.0,
            'total_gain_loss_percent': float(self.total_gain_loss_percent) if self.total_gain_loss_percent else 0.0,
            'last_updated': self.last_updated.isoformat() if self.last_updated else None,
            'holdings_count': self.holdings.count()
        }
    
    def __repr__(self):
        return f'<Portfolio User {self.user_id} - Value: {self.total_value}>'


class Holding(db.Model):
    __tablename__ = 'holdings'
    
    id = db.Column(db.Integer, primary_key=True)
    portfolio_id = db.Column(db.Integer, db.ForeignKey('portfolios.id', ondelete='CASCADE'), nullable=False, index=True)
    asset_type = db.Column(db.String(50), nullable=False)  # equity, debt, gold, international
    asset_name = db.Column(db.String(200), nullable=False)
    quantity = db.Column(db.Numeric(15, 4), nullable=False)
    current_price = db.Column(db.Numeric(15, 2), nullable=False)
    purchase_price = db.Column(db.Numeric(15, 2), nullable=False)
    allocation = db.Column(db.Numeric(5, 2), nullable=False)  # Percentage allocation
    ai_recommendation = db.Column(db.String(20), nullable=True)  # buy, hold, sell
    last_updated = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    def to_dict(self):
        """Convert holding to dictionary"""
        return {
            'id': self.id,
            'portfolio_id': self.portfolio_id,
            'asset_type': self.asset_type,
            'asset_name': self.asset_name,
            'quantity': float(self.quantity) if self.quantity else 0.0,
            'current_price': float(self.current_price) if self.current_price else 0.0,
            'purchase_price': float(self.purchase_price) if self.purchase_price else 0.0,
            'allocation': float(self.allocation) if self.allocation else 0.0,
            'ai_recommendation': self.ai_recommendation,
            'current_value': float(self.quantity * self.current_price) if self.quantity and self.current_price else 0.0,
            'gain_loss': float((self.current_price - self.purchase_price) * self.quantity) if self.quantity and self.current_price and self.purchase_price else 0.0,
            'gain_loss_percent': float(((self.current_price - self.purchase_price) / self.purchase_price * 100)) if self.purchase_price and self.purchase_price > 0 else 0.0,
            'last_updated': self.last_updated.isoformat() if self.last_updated else None
        }
    
    def __repr__(self):
        return f'<Holding {self.asset_name} - {self.quantity} units>'

